### ISSUE TITLE
Theme status pages (and dashboard?) by defining "base colors" in Appearance setting

### ISSUE THREAD

**Issue by pro-sumer** [louislam/uptime-kuma#4642]
### 📑 I have found these related issues/pull requests

Unable to find any related issue.

### 🏷️ Feature Request Type

Status-page

### 🔖 Feature description

It would be great if one can configure Uptime Kuma's "base colors" in the web interface, instead of manually overwriting many individual status page elements using CSS.

Bonus: also theme the dashboard this way.

### ✔️ Solution

Allow the user to define the hex/RGB values for the "base colors" in the Appearance section of Uptime Kuma's settings.

### ❓ Alternatives

A CSS solution that does not require overwriting many individual status page items would also be good, but unfortunately I don't know how to do that (if this is already possible, a Wiki entry on how to do it would be appreciated).

### 📝 Additional Context

N/A

**Comment by CommanderStorm** [comment]
Currently, our system works by defining colours in SCSS in [`src/assets/vars.scss`](https://github.com/louislam/uptime-kuma/blob/master/src/assets/vars.scss).
This is then used in most places (not quite all, PRs welcome)
The colours are then used at [`src/assets/app.scss`](https://github.com/louislam/uptime-kuma/blob/master/src/assets/app.scss) to override the defaults by [bootstrap 5](https://getbootstrap.com/docs/5.2/getting-started/introduction/).

The problem is that SCSS variables are not changeable at runtime (SCSS is compiled).
One option would be to change these to [CSS variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties).

I think changes like you did on https://status.prosumer.dev/ are somewhat out of scope.
If somebody has such special requirements, they will be reaching for the CSS guns anyway… => don't see a big payoff of spending lots of effort here..

**Comment by pro-sumer** [comment]
> The problem is that SCSS variables are not changeable at runtime (SCSS is compiled).

That's what I thought. Thanks for confirming.

> If somebody has such special requirements, they will be reaching for the CSS guns anyway…

Luckily Uptime Kuma allows custom CSS. Thank you for that option.

**Comment by chakflying** [comment]
We should consider moving from sass variables to css variables to allow customization in this way.

**Comment by tayflo** [comment]
I've just noticed that beats are now drawn by canvas (`.heartbeat-canvas`) instead of individual elements, and so it seems (but I might be missing something here) it's no longer possible to change their colors with custom CSS (it was possible in version 1.23.17).

**Comment by CommanderStorm** [comment]
Unfortunate that this broke your usage, we just did not think about this usecase.

You can still overide the colors via the CSS variables `--beat-empty-color`, ..
https://github.com/kurama/uptime-kuma/commit/8ad09ccd13091117ebaef35eaca50a49b453231d

**Comment by tayflo** [comment]
Very nice, thanks! I've been able to change the colors of the beats on the canvas using a syntax like that:

```css
:root {
  --bs-primary: #fff;
  --bs-warning: #fff;
  --bs-danger: #fff;
  --maintenance: #fff;
  --beat-empty-color: #fff;
}
```

It's also nice that those values apply broadly, even if sometimes it is required to change the RGB triple instead, for instance for:

```css
.bg-danger {
  background-color: rgba(var(--bs-danger-rgb),var(--bs-bg-opacity)) !important;
}
```

I'm not quite sure why this is necessary; in practice, I've just seen `--bs-bg-opacity` set to 1. But I guess this exists to allow for some kind of flexibility.

---

Even though, I've not been able to change the text color on the tooltip appearing on mouseover. The tooltip disappear dynamically so I've not been able to inspect the element.

And I guess it's unaffected by CSS variables because it's [using](https://github.com/kurama/uptime-kuma/blob/574d42ca51cd5ed7ad6893aa92ae180dcdae6ca3/src/components/Tooltip.vue#L152-L170) the [SCSS variables](https://github.com/kurama/uptime-kuma/blob/574d42ca51cd5ed7ad6893aa92ae180dcdae6ca3/src/assets/vars.scss).

It's the same for the [overall status icon](https://github.com/kurama/uptime-kuma/blob/574d42ca51cd5ed7ad6893aa92ae180dcdae6ca3/src/pages/StatusPage.vue#L1496-L1511) but this one is more easily inspectable.

**Comment by CommanderStorm** [comment]
Feel free to contribute a code change if you know a better version.

I have been wanting to reduce the extremely high number of custom CSS for a while, but have not come to it.
Changes to html and thus are not something that I consider public api since that is just way too hard to to controll.
If we can move most of those usecases to css variables, that would be much simpler.

**Comment by nojhan** [comment]
If that can motivate someone: the current green/red color scheme is impossible to read by people with the most frequent color blindness. Accessibility best practices recommend to at least have blue/yellow, or better: make it configurable.

**Comment by tayflo** [comment]
> If that can motivate someone: the current green/red color scheme is impossible to read by people with the most frequent color blindness. Accessibility best practices recommend to at least have blue/yellow, or better: make it configurable.

I agree that configurable colors are a good thing regarding accessibility and that it's possible that the default color scheme may be improved.

But I'm not sure about the current green/red color scheme been completely impossible to read for red-green colorblind people (deuteranomaly, protanomaly, protanopia, deuteranopia), because the green is very light and the red is less so, which I guess makes them somewhat differentiable.

Nonetheless, the contrast between the gray and the red seems to be not quite readable.

For instance, here is the result of some simulators (which are imperfect but can give a rough idea; here it's for quite severe color blindness I think):

<img width="389" height="371" alt="Image" src="https://github.com/user-attachments/assets/b2f0e898-2e80-4b89-8174-5036cc361f08" />

From [davidmathlogic.com/colorblind](https://davidmathlogic.com/colorblind/#%235CDD8B-%23D43543-%23848484-%23070A10)

<img width="1413" height="806" alt="Image" src="https://github.com/user-attachments/assets/d3ca54c1-2ebb-4218-a176-b21e38b6bda5" />

From [cvd-simulator.think.somethingorotherwhatever.com](https://cvd-simulator.think.somethingorotherwhatever.com/)

**Comment by CommanderStorm** [comment]
I am happy to have your pull request for more things in our css to use the variables @nojhan 

I am not affected by this issue, so won't invest engineering time into it.
If you do go fixing this, please consider making the PRs as atomic as possible, this way it is simpler for me to test.

**Comment by nojhan** [comment]
> Nonetheless, the contrast between the gray and the red seems to be not quite readable.

The simulation checks if one can discriminate colors (and it is known that you cannot design colors that fit all disabilities). Not if those colors carry the necessary semantics for a given user. I'm not an accessibility expert myself, so I think the best is just to follow advices from UX experts.


> I am not affected by this issue, so won't invest engineering time into it.

Thanks for your understanding, kindness and inclusivity. I'm not affected by JavaScript myself, but I'll see what I can learn to support those who are affected and don't code.

**Comment by SenseiHiraku** [comment]
Just wanted to mention that the [CuteKuma Theme](https://github.com/dermv/cute-kuma) also broke with the canvas update: https://github.com/dermv/cute-kuma/issues/12.
I’m sure many users would appreciate a more modern approach to the heartbeat bars, as the current one used by Uptime Kuma doesn’t really seem fitting and modification is now blocked by the canvas change.

Other status page vendors like [BetterStack](https://i.imgur.com/88t6bp2.png), [OneUptime](https://i.imgur.com/8KL81Cd.png) or [Instatus](https://status.wistia.com/) use an approach similar to CuteKuma, with a more unified look. 

Uptime Kuma is by far the best open-source status page in terms of features, so I’d love to see it become even better on the frontend side as well. I’m not suggesting that Uptime Kuma should change its design to match the others, but I would appreciate it if theming capabilities remained open to preserve user choice and keep themes like CuteKuma possible.

**Comment by CommanderStorm** [comment]
The change to the heartbeat bars was made because it is a 70% performance gain, so fixes this aspect a bit more.
That was a lot of work, we are not going back.

I think all aspects are confiigureble via CSS variables. Not non-breaking due to us not having realised we promised an API there, but I don't see a hood way to fix this.

**Comment by CommanderStorm** [comment]
> I'll see what I can learn to support those who are affected 

@nojhan thanks for looking into this. I am a bit overloaded on my real work right now, so cannot look into it myself. If there is a pr with screenshots though I can review it

**Comment by tayflo** [comment]
> @nojhan: The simulation checks if one can discriminate colors (and it is known that you cannot design colors that fit all disabilities). Not if those colors carry the necessary semantics for a given user.

(1) I think the semantics can be quite easily inferred, since in this case the text "up" and "down" serve as a legend (but yes, the text is a necessity to convey proper semantics here).

(2) There are color palettes designed especially to be foolproof for colorblindness, for instance, Viridis.

If you were to dig on the topic, you may look at the Web Content Accessibility Guidelines (WCAG): [version 2](https://www.w3.org/WAI/standards-guidelines/wcag/), and draft for [version 3](https://www.w3.org/WAI/standards-guidelines/wcag/wcag3-intro/).

**Comment by tayflo** [comment]
> sometimes it is required to change the RGB triple instead

On that matter, if it were to be still necessary to have transparent shades, one might use `color-mix` ([MDN](https://developer.mozilla.org/docs/Web/CSS/Reference/Values/color_value/color-mix)) in order not to declare each value twice, in hexadecimal and in RGB triple.

Example:

```css
--bs-danger: #ff0000;
--bs-bg-opacity: 0.2;  /* I think 20% should also work */
background-color: color-mix(in srgb, var(--bs-danger) var(--bs-bg-opacity), transparent);
```

Instead of:

```css
--bs-danger: #ff0000;        /* For HEX usage */
--bs-danger-rgb: 255, 0, 0;  /* Re-decleration of the color for RGB usage */
--bs-bg-opacity: 0.2;
background-color: rgba(var(--bs-danger-rgb), var(--bs-bg-opacity));
```
